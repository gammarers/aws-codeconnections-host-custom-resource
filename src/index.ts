// ã€€import * as crypto from 'crypto';
import { ResourceNaming } from '@gammarers/aws-resource-naming';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

// Bitbucket | GitHub | GitHubEnterpriseServer | GitLab | GitLabSelfManaged
export enum CodeConnectionsHostProviderType {
  BIT_BUCKET = 'Bitbucket',
  GIT_HUB = 'GitHub',
  GIT_HUB_ENTERPRISE_SERVER = 'GitHubEnterpriseServer',
  GIT_LAB = 'GitLab',
  GIT_LAB_SELF_MANAGED = 'GitLabSelfManaged',
}

export enum ResponseField {
  HOST_ARN = 'HostArn',
}

export interface ResourceNamingOptions {
  readonly naming: ResourceNaming.AutoNaming | ResourceNaming.DefaultNaming | {
    type: ResourceNaming.NamingType.CUSTOM;
    names: {
      functionName: string;
      functionRoleName: string;
    };
  };
}

export interface CodeConnectionsHostCustomResourceProps {
  readonly name: string;
  readonly providerEndpoint: string;
  readonly providerType: CodeConnectionsHostProviderType;
  readonly resouceNamingOption?: ResourceNamingOptions;
}

export class CodeConnectionsHostCustomResource extends cr.AwsCustomResource {

  constructor(scope: Construct, id: string, props: CodeConnectionsHostCustomResourceProps) {

    // ðŸ‘‡ Create random 8 length string
    const random = ResourceNaming.createRandomString(`${cdk.Names.uniqueId(scope)}.${props.name}.${props.providerEndpoint}.${props.providerType}`);
    const autoNaming = {
      functionName: `custom-resource-codeconnection-host-${random}-func`,
      functionRoleName: `custom-resource-codeconnection-host-${random}-func-exc-role`,
    };
    const naming = ResourceNaming.naming(autoNaming, props.resouceNamingOption);

    const account = cdk.Stack.of(scope).account;
    const region = cdk.Stack.of(scope).region;

    const functionRole = new iam.Role(scope, 'FunctionRole', {
      roleName: naming.names.functionRoleName,
      description: 'Custom Resource Function Execution Role.',
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        CustomResourcePolicy: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'codeconnections:CreateHost',
                'codeconnections:UpdateHost',
                'codeconnections:DeleteHost',
              ],
              resources: [
                `arn:aws:codeconnections:${region}:${account}:*`,
              ],
            }),
          ],
        }),
      },
    });

    super(scope, id, {
      functionName: naming.names.functionName,
      role: functionRole,
      timeout: cdk.Duration.seconds(15),
      installLatestAwsSdk: true,
      onCreate: {
        service: 'CodeConnections',
        action: 'createHost',
        parameters: {
          Name: props.name,
          ProviderEndpoint: props.providerEndpoint,
          ProviderType: props.providerType,
        },
        physicalResourceId: cr.PhysicalResourceId.fromResponse('HostArn'),
      },
      onUpdate: {
        service: 'CodeConnections',
        action: 'updateHost',
        parameters: {
          HostArn: new cr.PhysicalResourceIdReference(),
          ProviderEndpoint: props.providerEndpoint,
        },
      },
      onDelete: {
        service: 'CodeConnections',
        action: 'deleteHost',
        parameters: {
          HostArn: new cr.PhysicalResourceIdReference(),
        },
      },
    });
  }

  findHostArn(): string {
    return this.getResponseField(ResponseField.HOST_ARN);
  }

}