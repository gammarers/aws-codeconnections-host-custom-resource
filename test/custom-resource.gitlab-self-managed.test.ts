import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as codeconnections from 'aws-cdk-lib/aws-codeconnections';
import { AwsCustomResource } from 'aws-cdk-lib/custom-resources';
import { CodeConnectionsHostCustomResource, CodeConnectionsHostProviderType } from '../src';

describe('CustomResource Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const gitLabSelfManagedConnectionHostCustomResource = new CodeConnectionsHostCustomResource(stack, 'GitLabSelfManagedCodeConnectionsHostCustomResource', {
    name: 'gitlab.example.com',
    providerEndpoint: 'https://gitlab.example.com',
    providerType: CodeConnectionsHostProviderType.GIT_LAB_SELF_MANAGED,
  });

  const hostArn = gitLabSelfManagedConnectionHostCustomResource.getHostArn();

  new codeconnections.CfnConnection(stack, 'Connection', {
    connectionName: 'example-gitlab-self-managed-connection',
    hostArn,
  });

  const template = Template.fromStack(stack);

  it('Is GitLab Self Managed Connectrion Host CustomResource', () => {
    expect(gitLabSelfManagedConnectionHostCustomResource).toBeInstanceOf(AwsCustomResource);
  });

  it('Is hostArn string', () => {
    expect(hostArn).not.toBeNull();
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
