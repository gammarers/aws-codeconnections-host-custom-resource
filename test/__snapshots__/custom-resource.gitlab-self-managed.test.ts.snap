// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomResource Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "FunctionRole111A5701",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "97f30e67419a1676a2215492723e5add1aa491caf0cbe2dd878fc4fab0468cd4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "FunctionRole111A5701",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "Connection": {
      "Properties": {
        "ConnectionName": "example-gitlab-self-managed-connection",
        "HostArn": {
          "Fn::GetAtt": [
            "GitLabSelfManagedCodeConnectionsHostCustomResource31CD67FB",
            "HostArn",
          ],
        },
      },
      "Type": "AWS::CodeConnections::Connection",
    },
    "FunctionRole111A5701": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Custom Resource Function Execution Role.",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "codeconnections:CreateHost",
                    "codeconnections:UpdateHost",
                    "codeconnections:DeleteHost",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:codeconnections:us-east-1:123456789012:*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CustomResourcePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GitLabSelfManagedCodeConnectionsHostCustomResource31CD67FB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Create": "{"service":"CodeConnections","action":"createHost","parameters":{"Name":"gitlab.example.com","ProviderEndpoint":"https://gitlab.example.com","ProviderType":"GitLabSelfManaged"},"physicalResourceId":{"responsePath":"HostArn"}}",
        "Delete": "{"service":"CodeConnections","action":"deleteHost","parameters":{"HostArn":"PHYSICAL:RESOURCEID:"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{"service":"CodeConnections","action":"updateHost","parameters":{"HostArn":"PHYSICAL:RESOURCEID:","ProviderEndpoint":"https://gitlab.example.com"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
